File: mobile/src/main/java/uk/co/appsbystudio/geoshare/utils/geocoder/GeocodingFromLatLngTask.java
package uk.co.appsbystudio.geoshare.utils.geocoder;

import android.os.AsyncTask;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;

import uk.co.appsbystudio.geoshare.utils.json.UrlUtil;

public class GeocodingFromLatLngTask extends AsyncTask<Void, Void, String> {

    private final Double lat;
    private final Double lng;

    public GeocodingFromLatLngTask(Double lat, Double lng) {
        this.lat = lat;
        this.lng = lng;
    }

    @Override
    protected String doInBackground(Void... voids) {
        String finalAddress = null;

        String url = UrlUtil.getReverseGeocodingUrl(lat, lng);
        try {
            String jsonResponse = UrlUtil.downloadUrl(url);

            System.out.println(jsonResponse);

            JSONObject jsonObject = new JSONObject(jsonResponse);

            finalAddress = ((JSONArray) jsonObject.get("results")).getJSONObject(0).getString("formatted_address");

        } catch (IOException | JSONException e) {
            e.printStackTrace();
        }

        return finalAddress;
    }
}

File: mobile/src/main/java/uk/co/appsbystudio/geoshare/utils/geocoder/GeocodingFromLatLngTask.kt
package uk.co.appsbystudio.geoshare.utils.geocoder

import android.arch.lifecycle.MutableLiveData
import android.os.AsyncTask
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import uk.co.appsbystudio.geoshare.utils.downloadUrl
import uk.co.appsbystudio.geoshare.utils.getReverseGeocodingUrl
import java.io.IOException

class GeocodingFromLatLngTask(private val lat: Double, private val lng: Double, private val liveData: MutableLiveData<String>? = null) : AsyncTask<Void, Void, String>() {

    override fun doInBackground(vararg voids: Void): String? {
        var finalAddress: String? = null

        val url = getReverseGeocodingUrl(lat, lng)
        try {
            val jsonResponse = downloadUrl(url)

            val jsonObject = JSONObject(jsonResponse)

            finalAddress = (jsonObject.get("results") as JSONArray).getJSONObject(0).getString("formatted_address")

        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: JSONException) {
            e.printStackTrace()
        }

        return finalAddress
    }

    override fun onPostExecute(result: String?) {
        super.onPostExecute(result)

        if (liveData != null && result != null && !result.isBlank()) {
            liveData.value = result
            return
        }

        liveData?.value = "Could not get address..."
    }
}